src/core_plugins/kibana/public/context/api/context.js

```diff
@@ -49,20 +105,101 @@ function fetchContextProvider(courier, Private) {
       .set('version', true)
       .set('size', size)
       .set('filter', filters)
+      .set('searchAfter', [timeValue, tieBreakerValue])
+      .set('sort', [
+        { [timeField]: timeSortDirection },
+        { [tieBreakerField]: tieBreakerSortDirection },
+      ]);
+  }
+
+  async function performQuery(
+    searchSource,
+    timeField,
+    timeSortDirection,
+    fromTimeValue,
+    toTimeValue,
+    expectedSize,
+    maxTimeValue = null
+  ) {
+    const response = await searchSource
       .set('query', {
         query: {
-          match_all: {},
+          constant_score: {
+            filter: {
+              // match_all: {},
+              range: {
+                [timeField]: {
+                  [timeSortDirection === 'asc' ? 'gte' : 'lte']: fromTimeValue,
+                  [timeSortDirection === 'asc' ? 'lte' : 'gte']: toTimeValue,
+                }
+              },
+            },
+          },
         },
         language: 'lucene'
       })
-      .set('searchAfter', anchorDocument.sort)
-      .set('sort', sort);
+      .fetchAsRejectablePromise();
+
+    const hits = _.get(response, ['hits', 'hits'], []);
+    const nextToTimeValue = clamp(
+      toTimeValue + (toTimeValue - fromTimeValue),
+      0,
+      maxTimeValue !== null ? maxTimeValue : Number.POSITIVE_INFINITY
+    );
+
+    if (
+      hits.length >= expectedSize ||
+      (maxTimeValue !== null && toTimeValue >= maxTimeValue) ||
+      toTimeValue <= 0
+    ) {
+      return hits;
+    } else {
+      if (maxTimeValue === null) {
+        maxTimeValue = await getMaxTimeValue(searchSource.get('index'), timeField);
+      }
+      return await performQuery(
```

Comment:
Since the range query is now part of the filter context of a `constant_score` query, the caching argument should apply as it is. Whether the connection overhead of two smaller requests is more expensive than transferring a single larger one is hard to predict in general.

That said, I feel that determining the range first is appealing in an algorithmic sense and could improve the testability. I'll give it a try.

Reply:
> Whether the connection overhead of two smaller requests is more expensive than transferring a single larger one is hard to predict in general.

It’s not necessarily just 2 v 1 though. Currently we might request the same documents many times over, whereas I’m suggesting we only request them once.

Reply:
Hard to argue with that demo :+1: